pkg/store/configuration/configuration.go

```diff
@@ -258,42 +275,99 @@ func (s *configurationStore) List(ctx context.Context) ([]*configapi.Configurati
 
 func (s *configurationStore) Watch(ctx context.Context, ch chan<- configapi.ConfigurationEvent, opts ...WatchOption) error {
 	watchOpts := make([]_map.WatchOption, 0)
+	watcher := watcher{}
 	for _, opt := range opts {
+		switch e := opt.(type) {
+		case watchReplayOption:
+			watcher.replay = true
+		case watchIDOption:
+			watcher.watchID = e.id
+		}
 		watchOpts = opt.apply(watchOpts)
 	}
+	eventCh := make(chan configapi.ConfigurationEvent)
+	watcher.eventCh = eventCh
+	watcherID := uuid.New()
+	err := s.watchers.add(watcherID, watcher)
+	if err != nil {
+		return err
+	}
+	var configurations []configapi.Configuration
+	if watcher.replay {
+		s.cacheMu.RLock()
+		configurations = make([]configapi.Configuration, 0, len(s.cache))
+		for _, configuration := range s.cache {
+			configurations = append(configurations, configuration)
+
+		}
+		s.cacheMu.RUnlock()
 
-	mapCh := make(chan _map.Event)
-	if err := s.configurations.Watch(ctx, mapCh, watchOpts...); err != nil {
-		return errors.FromAtomix(err)
 	}
 
 	go func() {
 		defer close(ch)
-		for event := range mapCh {
-			if configuration, err := decodeConfiguration(event.Entry); err == nil {
-				var eventType configapi.ConfigurationEvent_EventType
-				switch event.Type {
-				case _map.EventReplay:
-					eventType = configapi.ConfigurationEvent_REPLAYED
-				case _map.EventInsert:
-					eventType = configapi.ConfigurationEvent_CREATED
-				case _map.EventRemove:
-					eventType = configapi.ConfigurationEvent_DELETED
-				case _map.EventUpdate:
-					eventType = configapi.ConfigurationEvent_UPDATED
-				default:
-					eventType = configapi.ConfigurationEvent_UPDATED
-				}
-				ch <- configapi.ConfigurationEvent{
-					Type:          eventType,
-					Configuration: *configuration,
-				}
+		for _, configuration := range configurations {
+			ch <- configapi.ConfigurationEvent{
+				Type:          configapi.ConfigurationEvent_REPLAYED,
+				Configuration: configuration,
 			}
 		}
+
+		for event := range eventCh {
+			ch <- event
+		}
 	}()
+
+	go func() {
+		<-ctx.Done()
+		err := s.watchers.remove(watcherID)
+		if err != nil {
+			return
+		}
+		close(ch)
```

Comment:
Actually, we should not close this channel here. Since this channel is dependent on an upstream channel (`eventCh`), it should be closed after the upstream channel is closed. Otherwise, `ch` can be closed before `eventCh` is read, and the for loop reading `eventCh` will try to write to a closed channel and cause a panic.